extensions/amp-analytics/0.1/visibility-helper.js

```diff
@@ -0,0 +1,396 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from '../../../src/log';
+
+
+/**
+ * A helper class that implements visibility calculations based on the
+ * visibility ratio. It's used for documents, embeds and individual element.
+ * @implements {../../../src/service.Disposable}
+ */
+export class VisibilityHelper {
+  /**
+   * @param {?VisibilityHelper} parent
+   * @param {!Object<string, *>} spec
+   * @param {number=} opt_iniVisibility
+   * @param {boolean=} opt_factorParent
+   */
+  constructor(parent, spec, opt_iniVisibility, opt_factorParent) {
```

Comment:
I don't quite understand why the "recursive" is needed here. let's talk offline.